src/backend/backend.h

```diff
@@ -37,20 +37,102 @@ namespace snmalloc
   template<
     SNMALLOC_CONCEPT(ConceptPAL) PAL,
     bool fixed_range,
-    typename PageMapEntry = MetaEntry>
+    template<typename> class PageMapEntry = MetaEntry>
   class BackendAllocator : public CommonConfig
   {
   public:
     using Pal = PAL;
 
+    /**
+     * This class will go away in the next PR and be replaced with:
+     *
+     * using SlabMetadata = Metaslab;
+     *
+     * In non-CHERI systems, the chunk field should be unnecessary.  It is
+     * present in this version so that we can make owning this state part of the
+     * back-end API contract in one commit and then removing its use in back
+     * ends that don't use it a separate one.
+     */
+    class SlabMetadata : public Metaslab
+    {
+      friend class BackendAllocator;
+      capptr::Chunk<void> chunk;
+    };
+
     class Pagemap
     {
       friend class BackendAllocator;
 
+    public:
+      /**
+       * Export the type stored in the pagemap.
+       * The following class could be replaced by:
+       *
+       * ```
+       * using Entry = PageMapEntry<SlabMetadata>;
+       * ```
+       *
+       * The full form here provides an example of how to extend the pagemap
+       * entries.  It also guarantees that the front end never directly
+       * constructs meta entries, it only ever reads them or modifies them in
+       * place.
+       */
+      class Entry : public PageMapEntry<SlabMetadata>
+      {
+        /**
+         * The private initialising constructor is usable only by this back end.
+         */
+        friend class BackendAllocator;
+
+        /**
+         * The private default constructor is usable only by the pagemap.
+         */
+        friend class FlatPagemap<MIN_CHUNK_BITS, Entry, PAL, fixed_range>;
+
+        /**
+         * The only constructor that creates newly initialised meta entries.
+         * This is callable only by the back end.  The front end may copy,
+         * query, and update these entries, but it may not create them
+         * directly.  This contract allows the back end to store any arbitrary
+         * metadata in meta entries when they are first constructed.
+         */
+        SNMALLOC_FAST_PATH
+        Entry(SlabMetadata* meta, uintptr_t ras)
+        : PageMapEntry<SlabMetadata>(meta, ras)
+        {}
+
+        /**
+         * Default constructor.  This must be callable from the pagemap.
+         */
+        SNMALLOC_FAST_PATH Entry() = default;
+
+        /**
+         * Copy assignment is used only by the pagemap.
+         */
+        Entry& operator=(const Entry& other)
+        {
+          PageMapEntry<SlabMetadata>::operator=(other);
+          return *this;
+        }
+      };
+
+    private:
```

Comment:
Could this be changed to protected instead? I promise to keep dependencies on it to a minimum. Also I think that get_metaentry_mut could be moved here.

Reply:
I assume that you'll implement your own backend, rather than subclassing this one.  If that isn't too much work, then we can remove some template parameters from this one.  I am not subclassing this for either of the backends in the process-sandboxing code.

Reply:
To clarify a bit, set_metaentry is ok to stay private as long as get_metaentry_mut is at least protected (it is public right now, which is good). It would help if concretePagemap was protected since I need access to the get_address method and a pointer to it to pass into a compartment.

Reply:
Without subclassing I would need to copy alloc_chunk, dealloc_chunk and alloc_meta_data as they are and continue copying it over as changes happen to those methods.

Reply:
If you want to subclass, then I'm inclined to remove the definition of `Entry` here.  It was intended as an example for you.  You're going to need to override `alloc_chunk` and `dealloc_chunk` if you want other things in `Entry` anyway, so the total shared code that you'll have with the superclass is a small amount of boilerplate.  I strongly suspect that this is goign to end up with more complexity.